crates/sui-core/src/execution_scheduler/execution_scheduler_impl.rs

```diff
@@ -199,57 +221,138 @@ impl ExecutionScheduler {
         };
         let _ = self.tx_ready_certificates.send(pending_cert);
     }
-}
 
-impl ExecutionSchedulerAPI for ExecutionScheduler {
-    fn enqueue(
+    fn schedule_balance_withdraws(
         &self,
-        certs: Vec<(Schedulable, ExecutionEnv)>,
+        certs: Vec<(VerifiedExecutableTransaction, SequenceNumber, ExecutionEnv)>,
         epoch_store: &Arc<AuthorityPerEpochStore>,
     ) {
-        // schedule all transactions immediately
-        let mut txns = Vec::with_capacity(certs.len());
-        let mut rest = Vec::new();
-
-        for (schedulable, env) in certs {
-            if let Schedulable::Transaction(tx) = schedulable {
-                txns.push((tx, env));
-            } else {
-                rest.push((schedulable, env));
+        if certs.is_empty() {
+            return;
+        }
+        let scheduler = self
+            .balance_withdraw_scheduler
+            .as_ref()
+            .expect("Balance withdraw scheduler must be enabled if there are withdraws");
+        let mut withdraws = BTreeMap::new();
+        let mut prev_version = None;
+        for (cert, version, _) in &certs {
+            let tx_withdraws = cert
+                .transaction_data()
+                .process_balance_withdraws()
+                .expect("Balance withdraws should have already been checked");
+            assert!(!tx_withdraws.is_empty());
+            if let Some(prev_version) = prev_version {
+                // Transactions must be in order.
+                assert!(prev_version <= *version);
             }
+            prev_version = Some(*version);
+            let tx_digest = *cert.digest();
+            withdraws
+                .entry(*version)
+                .or_insert(Vec::new())
+                .push(TxBalanceWithdraw {
+                    tx_digest,
+                    reservations: tx_withdraws,
+                });
         }
+        let mut receivers = BTreeMap::new();
+        for (version, tx_withdraws) in withdraws {
+            receivers.extend(scheduler.schedule_withdraws(version, tx_withdraws));
+        }
+        let scheduler = self.clone();
+        let epoch_store = epoch_store.clone();
+        spawn_monitored_task!(epoch_store.clone().within_alive_epoch(async move {
+            for (cert, _, env) in certs {
```

Comment:
i ran into a problem in the settlement txn PR where checkpoint construction halted because we were waiting on all txns to be ready instead of taking them one at a time. This happened any time we enqueue a batch of transactions that spans multiple PendingCheckpoints, which is any time that a randomness pending checkpoint is created.

(see fix here: https://github.com/MystenLabs/sui/pull/22380/files#diff-575a72e9e068f57f52f87b553f0819967f11020f1f6ca975022d40c1d2cc913fR274)

I expect there will be a similar problem here, and this should use FuturesUnordered as well so that it can always make progress

Reply:
Would you consider moving the scheduling into `enqueue_transactions` instead then? I still feel that in general things are easier to reason about that way (each transaction is already in their own thread)

Reply:
Updated to use FuturesUnordered for now